onnxruntime/contrib_ops/cuda/bert/skip_layer_norm_impl.cu

```diff
@@ -72,70 +82,170 @@ __global__ void SkipLayerNormKernel(
   // this must be true because n is the total size of the tensor
   assert(n % ld == 0);
-  const int grid_size = n / ld;
-
-  if (ld <= 32) {
-    constexpr int block_size = 32;
-    SkipLayerNormKernelSmall<T, block_size>
-        <<<grid_size, block_size, 0, stream>>>(ld, input, skip, beta, gamma, bias, epsilon, output);
-  } else if (ld <= 128) {
-    constexpr int block_size = 128;
-    SkipLayerNormKernelSmall<T, block_size>
-        <<<grid_size, block_size, 0, stream>>>(ld, input, skip, beta, gamma, bias, epsilon, output);
-  } else if (ld == 384) {
-    constexpr int block_size = 384;
-    SkipLayerNormKernelSmall<T, block_size>
-        <<<grid_size, block_size, 0, stream>>>(ld, input, skip, beta, gamma, bias, epsilon, output);
+  if (use_half2 && 0 == (n % 4)) {  // this will force float32 to use the original kernels.
+    const int grid_size = n / ld;
+    bool hasBias = (bias == nullptr) ? false : true;
+    if (ld <= 32) {
+      constexpr int block_size = 32 / 4;
+      SkipLayerNormKernelVec<T, block_size, 4>
+         <<<grid_size, block_size, 0, stream>>>(ld, input, skip, beta, gamma, bias, epsilon, output, hasBias);
+    } else if (ld == 64) {
+      constexpr int block_size = 64 / 4;
+      SkipLayerNormKernelVec<T, block_size, 4>
+          <<<grid_size, block_size, 0, stream>>>(ld, input, skip, beta, gamma, bias, epsilon, output, hasBias);
+    } else if (ld == 128) {
+      constexpr int block_size = 128 / 4;
+      SkipLayerNormKernelVec<T, block_size, 4>
+          <<<grid_size, block_size, 0, stream>>>(ld, input, skip, beta, gamma, bias, epsilon, output, hasBias);
+    } else if (ld == 384) {
+      constexpr int block_size = 384 / 4;
+      SkipLayerNormKernelVec<T, block_size, 4>
+          <<<grid_size, block_size, 0, stream>>>(ld, input, skip, beta, gamma, bias, epsilon, output, hasBias);
+    } else if (ld == 768) {
+      constexpr int block_size = 768 / 4;
+      SkipLayerNormKernelVec<T, block_size, 4>
+          <<<grid_size, block_size, 0, stream>>>(ld, input, skip, beta, gamma, bias, epsilon, output, hasBias);
+    } else if (ld == 1024) {
+      constexpr int block_size = 1024 / 4;
+      SkipLayerNormKernelVec<T, block_size, 4>
+          <<<grid_size, block_size, 0, stream>>>(ld, input, skip, beta, gamma, bias, epsilon, output, hasBias);
+    } else {
+      constexpr int block_size = 256;
+      SkipLayerNormKernel<T, block_size>
+          <<<grid_size, block_size, 0, stream>>>(ld, input, skip, beta, gamma, bias, epsilon, output);
+    }
   } else {
-    constexpr int block_size = 256;
-    SkipLayerNormKernel<T, block_size><<<grid_size, block_size, 0, stream>>>(ld, input, skip, beta, gamma, bias, epsilon, output);
+    const int grid_size = n / ld;
+    if (ld <= 32) {
+      constexpr int block_size = 32;
+      SkipLayerNormKernelSmall<T, block_size>
+          <<<grid_size, block_size, 0, stream>>>(ld, input, skip, beta, gamma, bias, epsilon, output);
+    } else if (ld <= 64) {
```

Comment:
Looks like the code in most branches are very similar, maybe we simply the branching logic here. Can we enable ILP=4 for all cases if ld%4==0?